### PR TITLE
Generic parser

### PR DESCRIPTION
**Pull request by ScrapCodes**

### CODE REVIEW

**Comment by keerath** [comment]
Scala Parser Combinators are slow... 
please refer this https://lihaoyi.github.io/fastparse/#Comparisons

**Comment by ScrapCodes** [comment]
Yes you are right. Partly, I am aware of its existence. Just did not had
the time to explore the new library. I hope the shift to fast parse is not
much change, almost the same constructs are supported. So we can do that
upgrade anytime.

Prashant Sharma

**Comment by ScrapCodes** [comment]
http://stackoverflow.com/questions/33384906/what-are-the-available-parsers-for-scala-programming-language/33385723#33385723 

Is equally interesting thread. Looks like someone has written a grammer for scala.

**Comment by ScrapCodes** [comment]
@amit-shar Can you take a look at the code, and upload indexes for scala repos as well ?

**Comment by amit-shar** [comment]
Looks good to me. Indexing of scala repos is in process.

File: core/src/main/scala/com/kodebeagle/spark/SparkIndexJobHelper.scala [hunk 1]
```diff
@@ -41,17 +41,18 @@ object SparkIndexJobHelper {
...
-  RDD[(List[(String, String)], Option[Repository], List[String], Statistics)] = {
-    val zipFileNameRDD = sc.binaryFiles(KodeBeagleConfig.githubDir).map { case (zipFile, stream) =>
+  def makeZipFileExtractedRDD(sc: SparkContext): RDD[(List[(String, String)],
+    List[(String, String)], Option[Repository], List[String], Statistics)] = {
+    val zipFileNameRDD = sc.binaryFiles(KodeBeagleConfig.githubDir).map {
+      case (zipFile, stream) =>
       (zipFile.stripPrefix("file:").stripPrefix("hdfs:"), stream)
     }
     val zipFileExtractedRDD = zipFileNameRDD.map { case (zipFileName, stream) =>
       val repo: Option[Repository] = RepoFileNameParser(zipFileName)
```
Comment (amit-shar) [inline core/src/main/scala/com/kodebeagle/spark/SparkIndexJobHelper.scala]: In 'RepoFileNameParser' class language of the repository is hard coded. Correction required.

**Comment by amit-shar** [comment]
NoSuchElementException
Complete Stacktrace 
 Exception in thread "main" org.apache.spark.SparkException: Job aborted due to stage failure: Task 10 in stage 0.0 failed 1 times, most recent failure: Lost task 10.0 in stage 0.0 (TID 10, localhost): java.util.NoSuchElementException: None.get
[error]     at scala.None$.get(Option.scala:347)
[error]     at scala.None$.get(Option.scala:345)
[error]     at com.kodebeagle.parser.ScalaParser$$anonfun$generateTokensWithBraceContext$1.apply(ScalaParser.scala:102)
[error]     at com.kodebeagle.parser.ScalaParser$$anonfun$generateTokensWithBraceContext$1.apply(ScalaParser.scala:95)
[error]     at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:252)
[error]     at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:252)
[error]     at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:221)
[error]     at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
[error]     at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
[error]     at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:252)
[error]     at scala.collection.AbstractTraversable.flatMap(Traversable.scala:104)
[error]     at com.kodebeagle.parser.ScalaParser$.generateTokensWithBraceContext(ScalaParser.scala:95)
[error]     at com.kodebeagle.spark.CreateIndexJobForMethods$$anonfun$main$1.apply(CreateIndexJobForMethods.scala:47)
[error]     at com.kodebeagle.spark.CreateIndexJobForMethods$$anonfun$main$1.apply(CreateIndexJobForMethods.scala:43)
[error]     at scala.collection.Iterator$$anon$11.next(Iterator.scala:370)
[error]     at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:396)
[error]     at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:369)
[error]     at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13$$anonfun$apply$6.apply$mcV$sp(PairRDDFunctions.scala:1108)
[error]     at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13$$anonfun$apply$6.apply(PairRDDFunctions.scala:1108)
[error]     at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13$$anonfun$apply$6.apply(PairRDDFunctions.scala:1108)
[error]     at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1285)
[error]     at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13.apply(PairRDDFunctions.scala:1116)
[error]     at org.apache.spark.rdd.PairRDDFunctions$$anonfun$saveAsHadoopDataset$1$$anonfun$13.apply(PairRDDFunctions.scala:1095)
[error]     at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
[error]     at org.apache.spark.scheduler.Task.run(Task.scala:70)
[error]     at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
[error]     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error]     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error]     at java.lang.Thread.run(Thread.java:745)

File: core/src/main/scala/com/kodebeagle/spark/SparkIndexJobHelper.scala (hunk 1 above)
Comment (ScrapCodes) [inline core/src/main/scala/com/kodebeagle/spark/SparkIndexJobHelper.scala]: Alright!

**Comment by ScrapCodes** [comment]
@amit-shar I have tried to address your comments. It should work fine now, give it another run ?

**Comment by amit-shar** [comment]
1. It is working fine now!
2. Uploaded indexes for scala repos locally. So that we can test the results.

**Pull request closed without merging**